slay/utils.py

```diff
@@ -205,80 +205,85 @@ def calc_mean_and_std_wf(
-
-    else:
-
-        mean_wf = cp.zeros(
-            (
-                n_clusters,
-                params["n_chan"],
-                params["pre_samples"] + params["post_samples"],
-            )
+        # recalculate mean_wf if it is not the right shape
+        if mean_wf.shape[0] == n_clusters:
+            try:
+                std_wf = np.load(std_wf_path)
+            except FileNotFoundError:
+                std_wf = np.array([])
+            if np.any(np.isnan(mean_wf)):
+                mean_wf = np.nan_to_num(mean_wf, nan=0)
+                # save the fixed mean waveform
+                np.save(mean_wf_path, mean_wf)
+
+            if return_spikes:
+                # Extracting spikes is faster than saving and loading them from file
+                for i in tqdm(cluster_ids, desc="Loading spikes"):
+                    spikes_i = extract_spikes(
+                        data,
+                        spike_times,
+                        i,
+                        params["pre_samples"],
+                        params["post_samples"],
+                        params["max_spikes"],
+                    )
+                    spikes[i] = spikes_i
+            return mean_wf, std_wf, spikes
+
+    bits_to_uV = get_bits_to_uV_factor(params)  # convert from bits to uV
+    bits_to_uV = cp.float32(bits_to_uV)  # convert to cupy float32
+    mean_wf = cp.zeros(
+        (
+            n_clusters,
+            params["n_chan"],
+            params["pre_samples"] + params["post_samples"],
+        )
+    )
+    std_wf = cp.zeros_like(mean_wf)
+    for i in tqdm(cluster_ids, desc="Calculating mean and std waveforms"):
+        spikes[i] = extract_spikes(
+            data,
+            spike_times,
+            i,
+            params["pre_samples"],
+            params["post_samples"],
+            params["max_spikes"],
         )
-        std_wf = cp.zeros_like(mean_wf)
-        for i in tqdm(cluster_ids, desc="Calculating mean and std waveforms"):
-            spikes[i] = extract_spikes(
-                data,
-                spike_times,
-                i,
-                params["pre_samples"],
-                params["post_samples"],
-                params["max_spikes"],
-            )
-            if len(spikes[i]) > 0:  # edge case
-                spikes_cp = cp.array(spikes[i])
-                mean_wf[i, :, :] = cp.mean(spikes_cp, axis=0)
-                std_wf[i, :, :] = cp.std(spikes_cp, axis=0)
-
-        tqdm.write("Saving mean and std waveforms...")
-        mean_wf *= get_bits_to_uV_factor(params)  # convert from bits to uV
-        cp.save(mean_wf_path, mean_wf)
-        cp.save(std_wf_path, std_wf)
-        # Convert to numpy arrays
-        mean_wf = cp.asnumpy(mean_wf)
-        std_wf = cp.asnumpy(std_wf)
+        if len(spikes[i]) > 0:  # edge case
+            spikes_cp = cp.array(spikes[i], dtype=cp.float32)
+            spikes_cp *= bits_to_uV # convert from bits to uV
```

Comment:
this might be too many multiplications to do? maybe we should convert to uV at the end outside the for loop? i think overflow shouldn't be a problem bc the mean and std functions multiply by 1/n before adding